Current version is not compatible with Raspberry Pi kernel 5.10.xx+

```
From https://github.com/pguyot/wm8960
 = [up to date]      tagtagtag-sound -> origin/tagtagtag-sound
 = [up to date]      master          -> origin/master
Already up to date.
make -C /lib/modules/5.10.11+/build M=/home/pi/wm8960 clean
make[1]: Entering directory '/usr/src/linux-headers-5.10.11+'
make[1]: Leaving directory '/usr/src/linux-headers-5.10.11+'
rm -f tagtagtag-mixerd tagtagtag-mixerd-test
cc -Wall -Werror -lasound -lpthread tagtagtag-mixerd.c -o tagtagtag-mixerd
make -C /lib/modules/5.10.11+/build M=/home/pi/wm8960 modules
make[1]: Entering directory '/usr/src/linux-headers-5.10.11+'
scripts/Makefile.lib:8: 'always' is deprecated. Please use 'always-y' instead
  CC [M]  /home/pi/wm8960/wm8960.o
/home/pi/wm8960/wm8960.c: In function ‘wm8960_configure_clocking’:
/home/pi/wm8960/wm8960.c:745:15: error: implicit declaration of function ‘snd_soc_component_read32’; did you mean ‘snd_soc_component_read’? [-Werror=implicit-function-declaration]
  u16 iface1 = snd_soc_component_read32(component, WM8960_IFACE1);
               ^~~~~~~~~~~~~~~~~~~~~~~~
               snd_soc_component_read
/home/pi/wm8960/wm8960.c:745:6: warning: unused variable ‘iface1’ [-Wunused-variable]
  u16 iface1 = snd_soc_component_read32(component, WM8960_IFACE1);
      ^~~~~~
/home/pi/wm8960/wm8960.c: At top level:
/home/pi/wm8960/wm8960.c:1318:3: error: ‘const struct snd_soc_dai_ops’ has no member named ‘digital_mute’
  .digital_mute = wm8960_mute,
   ^~~~~~~~~~~~
/home/pi/wm8960/wm8960.c:1318:18: error: initialization of ‘int (*)(struct snd_pcm_substream *, struct snd_soc_dai *)’ from incompatible pointer type ‘int (*)(struct snd_soc_dai *, int)’ [-Werror=incompatible-pointer-types]
  .digital_mute = wm8960_mute,
                  ^~~~~~~~~~~
/home/pi/wm8960/wm8960.c:1318:18: note: (near initialization for ‘wm8960_dai_ops.prepare’)
cc1: some warnings being treated as errors
make[2]: *** [scripts/Makefile.build:279: /home/pi/wm8960/wm8960.o] Error 1
make[1]: *** [Makefile:1808: /home/pi/wm8960] Error 2
make[1]: Leaving directory '/usr/src/linux-headers-5.10.11+'
make: *** [Makefile:17: all] Error 2
```

--- a/.github/workflows/arm-runner.yml
+++ b/.github/workflows/arm-runner.yml
@@ -1,0 +1,27 @@
+name: Checking driver compiles on various versions of raspbian/raspios
+on:
+  push:
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        base_image: ['raspbian_lite:2020-02-13', 'raspios_lite:2021-03-04']
+        cpu: [arm1176, cortex-a7]
+    steps:
+    - uses: actions/checkout@v2
+    - uses: pguyot/arm-runner-action@v1
+      with:
+        image_additional_mb: 1024
+        base_image: ${{ matrix.base_image }}
+        cpu: ${{ matrix.cpu }}
+        commands: |
+            kernel_version=`apt list --installed | grep raspberrypi-kernel | awk '{ print $2 }'`
+            wget -q http://archive.raspberrypi.org/debian/pool/main/r/raspberrypi-firmware/raspberrypi-kernel-headers_${kernel_version}_armhf.deb
+            dpkg --install raspberrypi-kernel-headers_${kernel_version}_armhf.deb
+            apt-get update -y
+            apt-get install -y libasound2-dev
+            ls /lib/modules/*/build -la
+            echo /usr/src/linux-headers*
+            make
+            make test

--- a/Makefile
+++ b/Makefile
@@ -14,10 +14,10 @@ kernel_img_gzip_offset := $(shell expr $(kernel_img_gzip_offset) + 20)
 kernel_version := $(shell dd if=/boot/kernel.img skip=$(kernel_img_gzip_offset) iflag=skip_bytes of=/dev/stdout | zgrep -aPom1 'Linux version \K\S+')
 
 all: tagtagtag-mixerd
-	make -C /lib/modules/$(kernel_version)/build M=$(PWD) modules
+	make -C /usr/src/linux-headers-$(kernel_version) M=$(PWD) modules
 
 clean:
-	make -C /lib/modules/$(kernel_version)/build M=$(PWD) clean
+	make -C /usr/src/linux-headers-$(kernel_version) M=$(PWD) clean
 	rm -f tagtagtag-mixerd tagtagtag-mixerd-test
 
 install: snd-soc-wm8960.ko snd-soc-max9759.ko tagtagtag-sound.dtbo tagtagtag-mixerd

--- a/tagtagtag-mixerd.c
+++ b/tagtagtag-mixerd.c
@@ -23,9 +23,9 @@
 #define CONFIG_FILE_PATH "/var/lib/tagtagtag-sound/mixer.conf"
 #define DEFAULT_TAG_SPEAKER_LOW         121
 #define DEFAULT_TAG_SPEAKER_HIGH        127
-#define DEFAULT_TAGTAG_SPEAKER_LOW      121
-#define DEFAULT_TAGTAG_SPEAKER_HIGH     127
-#define DEFAULT_SPEAKER_BASE            249
+#define DEFAULT_TAGTAG_SPEAKER_LOW      110
+#define DEFAULT_TAGTAG_SPEAKER_HIGH     120
+#define DEFAULT_SPEAKER_BASE            255
 #define DEFAULT_HEADPHONE_LOW           227
 #define DEFAULT_HEADPHONE_HIGH          249
 #define LINEOUT_MODE_LINEOUT            0
@@ -33,7 +33,7 @@
 #define DEFAULT_LINEOUT_MODE            LINEOUT_MODE_LINEOUT
 #define DEFAULT_MICROPHONE_ENABLED      1
 #define DEFAULT_INPUT1_BASE             3
-#define DEFAULT_CAPTURE_BASE            39
+#define DEFAULT_CAPTURE_BASE            90
 #define PID_FILE_PATH "/run/tagtagtag-mixerd.pid"
 
 struct config {

--- a/wm8960.c
+++ b/wm8960.c
@@ -742,7 +742,7 @@ static int wm8960_configure_clocking(struct snd_soc_component *component)
 {
 	struct wm8960_priv *wm8960 = snd_soc_component_get_drvdata(component);
 	int freq_out, freq_in;
-	u16 iface1 = snd_soc_component_read32(component, WM8960_IFACE1);
+	u16 iface1 = snd_soc_component_read(component, WM8960_IFACE1);
 	int i, j, k;
 	int ret;
 
@@ -806,7 +806,7 @@ static int wm8960_hw_params(struct snd_pcm_substream *substream,
 {
 	struct snd_soc_component *component = dai->component;
 	struct wm8960_priv *wm8960 = snd_soc_component_get_drvdata(component);
-	u16 iface = snd_soc_component_read32(component, WM8960_IFACE1) & 0xfff3;
+	u16 iface = snd_soc_component_read(component, WM8960_IFACE1) & 0xfff3;
 	bool tx = substream->stream == SNDRV_PCM_STREAM_PLAYBACK;
 	int i;
 
@@ -873,7 +873,7 @@ static int wm8960_hw_free(struct snd_pcm_substream *substream,
 	return 0;
 }
 
-static int wm8960_mute(struct snd_soc_dai *dai, int mute)
+static int wm8960_mute(struct snd_soc_dai *dai, int mute, int direction)
 {
 	struct snd_soc_component *component = dai->component;
 
@@ -888,7 +888,7 @@ static int wm8960_set_bias_level_out3(struct snd_soc_component *component,
 				      enum snd_soc_bias_level level)
 {
 	struct wm8960_priv *wm8960 = snd_soc_component_get_drvdata(component);
-	u16 pm2 = snd_soc_component_read32(component, WM8960_POWER2);
+	u16 pm2 = snd_soc_component_read(component, WM8960_POWER2);
 	int ret;
 
 	switch (level) {
@@ -978,7 +978,7 @@ static int wm8960_set_bias_level_capless(struct snd_soc_component *component,
 					 enum snd_soc_bias_level level)
 {
 	struct wm8960_priv *wm8960 = snd_soc_component_get_drvdata(component);
-	u16 pm2 = snd_soc_component_read32(component, WM8960_POWER2);
+	u16 pm2 = snd_soc_component_read(component, WM8960_POWER2);
 	int reg, ret;
 
 	switch (level) {
@@ -1197,7 +1197,7 @@ static int wm8960_set_pll(struct snd_soc_component *component,
 	if (!freq_in || !freq_out)
 		return 0;
 
-	reg = snd_soc_component_read32(component, WM8960_PLL1) & ~0x3f;
+	reg = snd_soc_component_read(component, WM8960_PLL1) & ~0x3f;
 	reg |= pll_div.pre_div << 4;
 	reg |= pll_div.n;
 
@@ -1240,23 +1240,23 @@ static int wm8960_set_dai_clkdiv(struct snd_soc_dai *codec_dai,
 
 	switch (div_id) {
 	case WM8960_SYSCLKDIV:
-		reg = snd_soc_component_read32(component, WM8960_CLOCK1) & 0x1f9;
+		reg = snd_soc_component_read(component, WM8960_CLOCK1) & 0x1f9;
 		snd_soc_component_write(component, WM8960_CLOCK1, reg | div);
 		break;
 	case WM8960_DACDIV:
-		reg = snd_soc_component_read32(component, WM8960_CLOCK1) & 0x1c7;
+		reg = snd_soc_component_read(component, WM8960_CLOCK1) & 0x1c7;
 		snd_soc_component_write(component, WM8960_CLOCK1, reg | div);
 		break;
 	case WM8960_OPCLKDIV:
-		reg = snd_soc_component_read32(component, WM8960_PLL1) & 0x03f;
+		reg = snd_soc_component_read(component, WM8960_PLL1) & 0x03f;
 		snd_soc_component_write(component, WM8960_PLL1, reg | div);
 		break;
 	case WM8960_DCLKDIV:
-		reg = snd_soc_component_read32(component, WM8960_CLOCK2) & 0x03f;
+		reg = snd_soc_component_read(component, WM8960_CLOCK2) & 0x03f;
 		snd_soc_component_write(component, WM8960_CLOCK2, reg | div);
 		break;
 	case WM8960_TOCLKSEL:
-		reg = snd_soc_component_read32(component, WM8960_ADDCTL1) & 0x1fd;
+		reg = snd_soc_component_read(component, WM8960_ADDCTL1) & 0x1fd;
 		snd_soc_component_write(component, WM8960_ADDCTL1, reg | div);
 		break;
 	default:
@@ -1315,7 +1315,7 @@ static int wm8960_set_dai_sysclk(struct snd_soc_dai *dai, int clk_id,
 static const struct snd_soc_dai_ops wm8960_dai_ops = {
 	.hw_params = wm8960_hw_params,
 	.hw_free = wm8960_hw_free,
-	.digital_mute = wm8960_mute,
+	.mute_stream = wm8960_mute,
 	.set_fmt = wm8960_set_dai_fmt,
 	.set_clkdiv = wm8960_set_dai_clkdiv,
 	.set_pll = wm8960_set_dai_pll,

--- a/wm8960.c
+++ b/wm8960.c
@@ -15,6 +15,7 @@
 #include <linux/clk.h>
 #include <linux/i2c.h>
 #include <linux/slab.h>
+#include <linux/version.h>
 #include <sound/core.h>
 #include <sound/pcm.h>
 #include <sound/pcm_params.h>
@@ -22,6 +23,12 @@
 #include <sound/initval.h>
 #include <sound/tlv.h>
 #include <sound/wm8960.h>
+
+#include <linux/version.h>
+
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5,0,0)
+#define snd_soc_component_read snd_soc_component_read32
+#endif
 
 #include "wm8960.h"
 
@@ -873,7 +880,11 @@ static int wm8960_hw_free(struct snd_pcm_substream *substream,
 	return 0;
 }
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5,0,0)
+static int wm8960_mute(struct snd_soc_dai *dai, int mute)
+#else
 static int wm8960_mute(struct snd_soc_dai *dai, int mute, int direction)
+#endif
 {
 	struct snd_soc_component *component = dai->component;
 
@@ -1315,7 +1326,11 @@ static int wm8960_set_dai_sysclk(struct snd_soc_dai *dai, int clk_id,
 static const struct snd_soc_dai_ops wm8960_dai_ops = {
 	.hw_params = wm8960_hw_params,
 	.hw_free = wm8960_hw_free,
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5,0,0)
+	.digital_mute = wm8960_mute,
+#else
 	.mute_stream = wm8960_mute,
+#endif
 	.set_fmt = wm8960_set_dai_fmt,
 	.set_clkdiv = wm8960_set_dai_clkdiv,
 	.set_pll = wm8960_set_dai_pll,

